### PR TITLE
DB-9068 Use Java serialization for user defined aggregates

### PR DESCRIPTION
**Pull request by dgomezferro**
It seems we are not serializing these objects to disk, so I wasn't able to test the upgrade logic. If you can come up with an scenario where we'd store them on disk let me know.

### CODE REVIEW

**Comment by dgomezferro** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: splice_machine/src/main/java/com/splicemachine/tools/StringConcat.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    @Override
+    public void merge(StringConcat otherAggregator) {
+        agg.append(otherAggregator.agg);
+    }
+
+    @Override
+    public String terminate() {
+        return agg.toString();
+    }
+}
```
Comment (jyuanca) [inline splice_machine/src/main/java/com/splicemachine/tools/StringConcat.java]: Minor issue - license header is missing

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java [hunk 2]
```diff
@@ -63,6 +67,7 @@
     ///////////////////////////////////////////////////////////////////////////////////
 
     private static final int FIRST_VERSION = 0;
+	private static final int SECOND_VERSION = 1; // use java serialization for _aggregator since it will not be registered in our Kryo registry
```
Comment (arnaud-splice) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java]: [minor] replace tabs with spaces

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java [hunk 3]
```diff
@@ -180,9 +185,15 @@ public ExecAggregator newAggregator()
 	 */
 	public void writeExternal(ObjectOutput out) throws IOException
 	{
-		out.writeInt( FIRST_VERSION );
-        
-        out.writeObject( _aggregator );
+		out.writeInt( SECOND_VERSION );
```
Comment (arnaud-splice) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java]: [minor] replace tabs with spaces

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java (hunk 2 above)
Comment (arnaud-splice) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java]: [minor] replace tabs with spaces

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java [hunk 4]
```diff
@@ -195,9 +206,18 @@ public void writeExternal(ObjectOutput out) throws IOException
 	public void readExternal(ObjectInput in) 
 		throws IOException, ClassNotFoundException
 	{
-		in.readInt();   // unused until we have a second rev of this class
-
-        _aggregator = (Aggregator) in.readObject();
+		int v = in.readInt();
```
Comment (arnaud-splice) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java]: [minor] replace tabs with spaces

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/SpliceUDTIT.java [hunk 5]
```diff
@@ -202,4 +205,23 @@ public void testConnection() throws Exception {
...
+        methodWatcher.execute("create derby aggregate string_concat for varchar(2000) external name 'com.splicemachine.tools.StringConcat'");
+
+        methodWatcher.execute("create table strings (v varchar(20))");
+        methodWatcher.execute("insert into strings values 'a','b','c','d'");
+
+        for (boolean useSpark : Arrays.asList(true, false)) {
+            ResultSet rs = methodWatcher.executeQuery("select string_concat(v) from strings --splice-properties useSpark="+useSpark);
+            Assert.assertTrue(rs.next());
+            String res = rs.getString(1);
+            Assert.assertTrue(res.contains("a"));
```
Comment (arnaud-splice) [inline splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/SpliceUDTIT.java]: `assertThat(res, allOf(containsString("a"), containsString("b"), containsString("c"), containsString("d")))`
would bring a better error message.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java (hunk 4 above)
Comment (dgomezferro) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/UserDefinedAggregator.java]: This file has mixed indenting, both options would look bad. I'd rather fix this with a repository wide reformatting effort.

**Comment by dgomezferro** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**